Support correlated scalar filter scan subqueries

kokosing opened the pull request:
Support correlated scalar filter scan subqueries

kokosing commented:
This fixes #8421  and probably some others as well as I noticed many people were asking about support for such support for such subqueries.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarFilterScanToJoin.java:
```diff
@@ -0,0 +1,94 @@
+import static com.facebook.presto.util.MorePredicates.isInstanceOfAny;
+import static java.util.Objects.requireNonNull;
+
+public class TransformCorrelatedScalarFilterScanToJoin
```
Can you add java doc please to explain what this does?  (and add that to the commit message too)

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -236,18 +236,18 @@ public PlanNode visitWindow(WindowNode node, RewriteContext<Void> context)
+            return canonicalize(node, context);
+        }
+
+        // once all the mappin function are ported to SymbolMapper, this method should be called from visitPlan
```
typo mappin

kokosing commented:
@rschlussel I addressed all your comments, I added javadoc and tests so I hope now it will be much easier to understand what it does. Would you like to do a deeper review?

Unfortunately, there are some travis failures, I will take a look at them tomorrow.

KBP-TDC commented:
Any user docs needed?

rschlussel commented:
Thanks! I'll look at it now.

kokosing commented:
@martint can you please take a look at the last commit? I curious about your opinion about this rewrite. I was also thinking about an hidden aggregation function which could work in the same way as enforce single row works. That way I would not need this case expression at the top of aggregation.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarFilterScanToJoin.java:
```diff
@@ -0,0 +1,124 @@
+                .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class))
+                .findAll();
+
+        if (filterNodes.size() != 1) {
```
I'm thinking about cases where filterNodes would be > 1. Do we ever do something like:
```
Filter: (x+1 > 3)
     Project: output: ( x+1, y)
             Filter(y < 7)
                    TableScan: output: x,y
```
Because in that case it would be valuable to pull the predicate up to combine the filter nodes so we could apply this rewrite (not suggesting we do that now, just thinking about limitations)

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarFilterScanToJoin.java:
```diff
@@ -0,0 +1,124 @@
+    private Optional<FilterNode> getFilterScan(PlanNode planNode, Lookup lookup)
+    {
+        if (!searchFrom(planNode, lookup)
+                .where(TableScanNode.class::isInstance)
```
Why do we need to be above a table scan?

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarSubqueryToJoinRewriter.java:
```diff
@@ -50,7 +50,7 @@
 import static java.util.Objects.requireNonNull;
 
 // TODO: move this class to TransformCorrelatedScalarAggregationToJoin when old optimizer is gone
```
Update the todo.  it seems like you want to continue to use some of this as a utility class for other subquery rewrites

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarSubqueryToJoinRewriter.java:
```diff
@@ -69,6 +81,58 @@ public ScalarSubqueryToJoinRewriter(FunctionRegistry functionRegistry, SymbolAll
         this.planNodeDecorrelator = new PlanNodeDecorrelator(idAllocator, lookup);
     }
 
+    public Optional<PlanNode> rewriteScalarFilterScan(LateralJoinNode lateralJoinNode, FilterNode filterNode)
```
Can you move this to TransformCorrelatedScalarFilterToJoin and make the relevant utility methods public instead?

kokosing commented:
@KBP-TDC 

> Any user docs needed?

Docs say:

> Support for correlated subqueries is limited. Not every standard form is supported.

Limitation still apply, but now more will be available.

kokosing commented:
@rschlussel I addressed all your comments, but the one with lateral type. I need to fix it as separate PR.

kokosing commented:
@rschlussel Regarding `LATERAL` fortunately I was wrong, everything is ok. We do not care lateral type as long we know that subquery is scalar. Either for `INNER` or `LEFT` results are the same.

rschlussel commented:
thanks for the speedy turnaround!  The code’s looking good, but I still don’t 100% understand the rewrite.  Could you give an example with a specific sql query (i.e. with the correlated bit spelled out), and what the plan starts as and what it changes to.   Or maybe I’ll step through it with a debugger on Monday.  
Also, could you answer my question about why this needs to be above a tablescan.  We don’t touch the table scan node, so it seems like it shouldn’t matter where the rows come from.

kokosing commented:
@rschlussel 
> Also, could you answer my question about why this needs to be above a tablescan. We don’t touch the table scan node, so it seems like it shouldn’t matter where the rows come from.

It does not matter. I wanted to start with a minimal version. In order to support more plan nodes I need to make `SymbolMapper` to be able to support more nodes to rewrite symbols them. I am going to change this to not have this limitation.

> Could you give an example with a specific sql query

Please take the a look at unit tests.

kokosing commented:
@sopel39 Can you please take a look?

martint commented:
> @martint can you please take a look at the last commit? I curious about your opinion about this rewrite. I was also thinking about an hidden aggregation function which could work in the same way as enforce single row works. That way I would not need this case expression at the top of aggregation.

I need to analyze the logic of the rewrite, but regarding the question about the aggregation, that makes sense.

kokosing commented:
I switched to MarkDistnictNode (as sopel39 suggested), this way thwre is no more aggregation and it much simplier.

kokosing commented:
@martint @sopel39 @findepi ping. Any review is appreciated.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -240,8 +240,10 @@ public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> contex
+            if (!SymbolsExtractor.extractUnique(context.get()).contains(node.getMarkerSymbol())) {
+                return context.defaultRewrite(node, context.get());
+            }
+            return new FilterNode(idAllocator.getNextId(), node, combineConjuncts(context.get()));
```
node should be default rewritten with `TRUE_LITERAL`.

sopel39 replied:
no need for `combineConjucts`.

findepi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -240,8 +240,10 @@ public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> contex
         {
-            checkState(!SymbolsExtractor.extractUnique(context.get()).contains(node.getMarkerSymbol()), "predicate depends on marker symbol");
-            return context.defaultRewrite(node, context.get());
+            if (!SymbolsExtractor.extractUnique(context.get()).contains(node.getMarkerSymbol())) {
```
> distinct marker used in predicate

I understand this broader set of working queries, perhaps new test would be good.

kokosing replied:
Not really. Typically `MarkDistinct` is used to provide distinct source values (let say column `A`)  to aggregate function. In order push down predicate on column `A` over `MarkDistinct` you would need to push down it over aggregation first, so that would mean that column `A` would must to survive the aggregation which is not possible. 
See: `SELECT count(DISTINCT name), count(DISTINCT regionkey) FROM region HAVING regionkey < 3;`

Filtered aggregations are not yet supported for aggregation with `DISTINCT`. Though I would expect them to filter column before marking distinct values.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedNonAggregationScalarToJoin.java:
```diff
@@ -0,0 +1,227 @@
+                        project(
+                                ImmutableMap.of("a2", expression("CASE (non_null) WHEN TRUE THEN (a * corr) ELSE null END")),
+                                filter(
+                                        "CASE is_distinct WHEN true THEN true ELSE CAST(fail(28, 'Scalar sub-query has returned multiple rows') AS BOOLEAN) END",
```
Same here

kokosing commented:
Rebased.

@sopel39 @martint @mbasmanova May kindly ask for a review? It is getting a bit old.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedNonAggregationScalarToJoin.java:
```diff
@@ -0,0 +1,181 @@
+ * <p>
+ * From:
+ * <pre>
+ * - LateralJoin (with correlation list: [C])
```
What if somebody specifically written lateral join and expects more than one result row from subquery?

Shouldn't scalar subqueries be represented by `ApplyNode`? Then you know that you expect at most one row from subquery.

Then the whole functionality could be expressed by two rules:
```
1. 
ApplyNode(correlation)
 - input
 - subquery

rewrite to:

LateralJoin[LEFT](correlation)
 - input
 - subquery

```
this rewrite fires only when `QueryCardinalityUtil#isAtMostScalar(subquery)==true`

the second rule
```
2.
ApplyNode(correlation)
 - input
 - subquery

rewrite to:

Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
  MarkDistinct(isDistinct)
    LateralJoin[LEFT](correlation)
     - AssignUniqueId
       - input
     - subquery

```
this rewrite fires when `1)` did not fire or `QueryCardinalityUtil#isAtMostScalar(subquery)==false`

`LateralJoin` rewrite would be taken care by other rules.
Am I missing something?

kokosing replied:
> What if somebody specifically written lateral join and expects more than one result row from subquery?

Sure, it should work. Here, subquery in lateral will be wrapped with `EnsureSingleRow`.

> Shouldn't scalar subqueries be represented by ApplyNode? Then you know that you expect at most one row from subquery.

I thought it works that way. Anyway, I still know it is scalar subquery as it is wrapped with `EnforceSingleRow`

What do you propose sounds reasonable, but this is kind of refactor which in my opinion exceeds the scope of this PR.

sopel39 replied:
> Here, subquery in lateral will be wrapped with `EnsureSingleRow`.

Still those two proposed rewrites apply. You just check if child of `EnsureSingleRow` is `isAtMostScalar`. If so, then taking out `EnsureSingleRow` should be enough. Otherwise you rewrite like in 2) (without `EnsureSingleRow` of course)

sopel39 replied:
We don't need to rewrite directly to join, do we? We can rewrite to lateral join:
```
LateralJoin(correlation, type)
 - input
 - subquery

rewrite to:

Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
  MarkDistinct(isDistinct)
    LateralJoin[type](correlation)
     - AssignUniqueId
       - input
     - subquery
```

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -236,8 +236,10 @@ public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> contex
         @Override
```
```
you would need to push down it
over aggregation first, so that would mean that column A would must to
survive the aggregation which is not possible.
```

You can pushdown predicates over aggregation if predicates are on columns that are part of each grouping set

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -295,6 +296,7 @@ public PlanOptimizers(
                         ImmutableSet.of(
                                 new RemoveUnreferencedScalarApplyNodes(),
                                 new TransformCorrelatedInPredicateToJoin(), // must be run after PruneUnreferencedOutputs
+                                new TransformCorrelatedNonAggregationScalarToJoin(metadata.getFunctionRegistry()),
```
inconsistent naming: the other rule (TransformCorrelatedNoAggregationSubqueryToJoin) says `No(Aggregation`, but this one says `Non(Aggregation)`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -236,8 +236,10 @@ public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> contex
         {
-            checkState(!SymbolsExtractor.extractUnique(context.get()).contains(node.getMarkerSymbol()), "predicate depends on marker symbol");
-            return context.defaultRewrite(node, context.get());
+            if (!SymbolsExtractor.extractUnique(context.get()).contains(node.getMarkerSymbol())) {
```
Predicates can be pushed only if they contain symbols from `MarkDistinctNode#distinctSymbols`.

mbasmanova commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6163,6 +6163,93 @@ public void testCorrelatedScalarSubqueries()
+
+        // multiple subquery output projections
+        assertQueryFails(
+                "SELECT name FROM nation n WHERE 'AFRICA' = (SELECT 'bleh' FROM region WHERE regionkey > n.regionkey)",
```
@kokosing Grzegorz, I can't figure out why this query reports a different error then the next one. Would you explain? I'd expect this one to fail at runtime, not at planning time.

mbasmanova commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6163,6 +6163,93 @@ public void testCorrelatedScalarSubqueries()
+
+        // multiple subquery output projections
+        assertQueryFails(
+                "SELECT name FROM nation n WHERE 'AFRICA' = (SELECT 'bleh' FROM region WHERE regionkey > n.regionkey)",
```
this is the same query as in L6173

mbasmanova commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6163,6 +6163,93 @@ public void testCorrelatedScalarSubqueries()
+                "SELECT (SELECT o.orderkey WHERE o.orderkey = 1) FROM orders o ORDER BY orderkey LIMIT 5",
+                UNSUPPORTED_CORRELATED_SUBQUERY_ERROR_MSG);
+        assertQuery(
+                "SELECT (SELECT o.orderkey * 2 WHERE o.orderkey = 1) FROM orders o ORDER BY orderkey LIMIT 5",
```
Why is this query passing while previous (seemingly simpler one) fails?

mbasmanova commented:
@kokosing Grzegorz, some commit messages have typos and could benefit from some editing. Consider the following rewrites.

for "Support PredicatePushdown with distinct marker used in predicate"

Typically MarkDistinct is used to provide distinct source
values, say column A, to aggregate functions. In order to push down
predicate on column A over MarkDistinct we need to push it down
over the aggregation first, so that means column A must survive the 
aggregation which is not possible.

However, there may be optimizer rules that use MarkDistinct
without aggregate functions, in those cases column A survives and it
is accessible to other nodes. This commit adds support for this case in
PredicatePushdown.

for "Rename variable"

The variable "source" in this context was not descriptive enough as it
was not clear whether it is a source of an aggregation or a lateral.

for "Prepare sources and create the join node in the same method"

It was not clear that the sources of the join node are prepared in one method
and then they are passed to the method that created the other source.

To make this more readable and clear this commit merges both into a single method,
which is now responsible for creating both the sources and the join node.
